reconcile/endpoints_discovery/integration.py

```diff
@@ -0,0 +1,347 @@
+            if integration_is_enabled(self.name, ns.cluster)
+            and (not cluster_names or ns.cluster.name in cluster_names)
+            and (not namespace_name or ns.name == namespace_name)
+            and not ns.delete
+        ]
+
+    def get_routes(self, oc_map: OCMap, namespace: NamespaceV1) -> list[Route]:
+        """Return the routes for the given namespace."""
+        oc = oc_map.get_cluster(namespace.cluster.name)
+        if not oc.project_exists(namespace.name):
+            logging.info(
+                f"{namespace.cluster.name}/{namespace.name}: Namespace does not exist (yet). Skipping for now!"
+            )
+            return []
+
+        return [
+            Route(
+                name=item["metadata"]["name"],
+                host=item["spec"]["host"],
+                tls=bool(item["spec"].get("tls")),
+            )
+            for item in oc.get_items(kind="Route", namespace=namespace.name)
+        ]
+
+    def get_endpoint_changes(
+        self,
+        app: str,
+        endpoint_prefix: str,
+        endpoint_template: str,
+        endpoints: Iterable[AppEndPointsV1],
+        routes: Iterable[Route],
+    ) -> tuple[EndpointsToAdd, EndpointsToChange, EndpointsToDelete]:
+        """Get all new/changed/deleted endpoints for the given namespace."""
+        if not routes and not endpoints:
+            # nothing to do
+            return [], [], []
+
+        diff = diff_any_iterables(
+            current=endpoints or [],
+            desired=routes,
+            # names are unique, so we can use them as keys
+            current_key=lambda endpoint: endpoint.name,
+            desired_key=lambda route: compile_endpoint_name(endpoint_prefix, route),
+            # compare the endpoint and route by name and url.
+            # we can't use other endpoint attributes because we don't want to query them.
+            # there is a note about that behavior in the template.
+            equal=lambda endpoint, route: endpoint.name
+            == compile_endpoint_name(endpoint_prefix, route)
+            and endpoint.url == route.url,
+        )
+
+        endpoints_to_add = []
+        endpoints_to_change = []
+        endpoints_to_delete = []
+
+        for add in diff.add.values():
+            logging.info(f"{app}: Adding endpoint for route {add.name}")
+            endpoints_to_add.append(
+                Endpoint(
+                    name=compile_endpoint_name(endpoint_prefix, add),
+                    data=render_template(
+                        endpoint_template,
+                        endpoint_name=compile_endpoint_name(endpoint_prefix, add),
+                        route=add,
+                    ),
+                )
+            )
+
+        for pair in diff.change.values():
+            if not pair.current.name.startswith(endpoint_prefix):
+                # do not update endpoints that were not created by this integration or for another namespace
```

Comment:
now that we fully manage all auto discovered endpoints, should we use filtered endpoints as current before diff to avoid duplicate startswith check? Like

```python
    current_endpoints = [e for e in endpoints if e.name.startswith(endpoint_prefix)]
```

Reply:
Done in [fd945f9](https://github.com/app-sre/qontract-reconcile/pull/4568/commits/fd945f96ad762c97fef670f14284d4d3f165fb7e)